// ngx-perfect-scrollbar/dist/lib/index.js
"use strict";
var perfect_scrollbar_component_1 = require("./perfect-scrollbar.component");
exports.PerfectScrollbarComponent = perfect_scrollbar_component_1.PerfectScrollbarComponent;
var perfect_scrollbar_directive_1 = require("./perfect-scrollbar.directive");
exports.PerfectScrollbarDirective = perfect_scrollbar_directive_1.PerfectScrollbarDirective;
var perfect_scrollbar_interfaces_1 = require("./perfect-scrollbar.interfaces");
exports.PerfectScrollbarConfig = perfect_scrollbar_interfaces_1.PerfectScrollbarConfig;
var perfect_scrollbar_module_1 = require("./perfect-scrollbar.module");
exports.provideDefaultConfig = perfect_scrollbar_module_1.provideDefaultConfig;
exports.provideForRootGuard = perfect_scrollbar_module_1.provideForRootGuard;
exports.PERFECT_SCROLLBAR_GUARD = perfect_scrollbar_module_1.PERFECT_SCROLLBAR_GUARD;
exports.PERFECT_SCROLLBAR_CONFIG = perfect_scrollbar_module_1.PERFECT_SCROLLBAR_CONFIG;
exports.PerfectScrollbarModule = perfect_scrollbar_module_1.PerfectScrollbarModule;
